File: cpp/classic/back_tracking/Arrangement.cpp
#include <iostream>
#include <vector>
using namespace std;

class Arrangement {
public:
    vector<vector<int>> solution(vector<int> &nums, int k) {
        vector<vector<int>> result(0);
        search(nums, k, 0, result);
        return result;
    }

private:
    void search(vector<int> &nums, int k, int start, vector<vector<int>> &result) {
        if (k == start) {
            result.emplace_back(vector<int>(nums.begin(), nums.begin() + k));
            return;
        }

        search(nums, k, start + 1, result);
        for (int i = start + 1; i < (int) nums.size(); ++i) {
            swap(nums[start], nums[i]);
            search(nums, k, start + 1, result);
            swap(nums[start], nums[i]);
        }
    }
};

File: java/classic/back_tracking/Arrangement.java
package classic.back_tracking;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Arrangement {
    public int[][] solution(int[] nums, int k) {
        List<int[]> result = new ArrayList<>();
        search(nums, k, 0, result);
        return result.toArray(new int[0][]);
    }

    private void search(int[] nums, int k, int start, List<int[]> result) {
        if (k == start) {
            result.add(Arrays.copyOf(nums, k));
            return;
        }

        search(nums, k, start + 1, result);
        for (int i = start + 1; i < nums.length; ++i) {
            swap(nums, start, i);
            search(nums, k, start + 1, result);
            swap(nums, start, i);
        }
    }

    private void swap(int[] nums, int i, int j) {
        nums[i] ^= nums[j];
        nums[j] ^= nums[i];
        nums[i] ^= nums[j];
    }
}
